pkg/domain/resourcegroup/runaway.go

```diff
@@ -596,25 +569,43 @@ func (r *RunawayChecker) CheckCopRespError(err error) error {
 }
 
 func (r *RunawayChecker) markQuarantine(now *time.Time) {
-	if r.setting.Watch == nil {
+	if r.settings == nil || r.settings.Watch == nil {
 		return
 	}
-	ttl := time.Duration(r.setting.Watch.LastingDurationMs) * time.Millisecond
+	ttl := time.Duration(r.settings.Watch.LastingDurationMs) * time.Millisecond
+
+	r.manager.markQuarantine(r.resourceGroupName, r.getSettingConvictIdentifier(), r.settings.Watch.Type, r.settings.Action, ttl, now)
+}
+
+func (r *RunawayChecker) markRunawayByIdentify(action rmpb.RunawayAction, now *time.Time) bool {
+	swapped := r.markedByRule.CompareAndSwap(false, true)
+	if swapped {
+		r.markRunaway("identify", action, now)
+		if !r.markedByWatch {
+			r.markQuarantine(now)
+		}
+	}
+	return swapped
+}
 
-	r.manager.markQuarantine(r.resourceGroupName, r.getSettingConvictIdentifier(), r.setting.Watch.Type, r.setting.Action, ttl, now)
+func (r *RunawayChecker) markRunawayByWatch(action rmpb.RunawayAction) {
+	r.markedByWatch = true
+	r.watchAction = action
+	now := time.Now()
+	r.markRunaway("watch", action, &now)
 }
 
-func (r *RunawayChecker) markRunaway(matchType RunawayMatchType, action rmpb.RunawayAction, now *time.Time) {
-	actionStr := strings.ToLower(rmpb.RunawayAction_name[int32(action)])
-	metrics.RunawayCheckerCounter.WithLabelValues(r.resourceGroupName, matchType.String(), actionStr).Inc()
+func (r *RunawayChecker) markRunaway(matchType string, action rmpb.RunawayAction, now *time.Time) {
+	actionStr := strings.ToLower(action.String())
```

Comment:
please change this line as well :)
https://github.com/pingcap/tidb/blob/7291829726b1c112e1179469c679b3bcd684eabd/pkg/executor/infoschema_reader.go#L3626